libs/content/workspace/src/tab/svg_editor/renderer.rs

```diff
@@ -0,0 +1,268 @@
+use std::collections::HashMap;
+
+use glam::f64::DVec2;
+use lyon::math::Point;
+use lyon::path::{AttributeIndex, LineCap, LineJoin};
+use lyon::tessellation::{
+    self, BuffersBuilder, FillVertexConstructor, StrokeOptions, StrokeTessellator,
+    StrokeVertexConstructor, VertexBuffers,
+};
+
+use rayon::prelude::*;
+use resvg::usvg::{ImageKind, Transform};
+
+use super::parser::{self, DiffState};
+use super::Buffer;
+
+const STROKE_WIDTH: AttributeIndex = 0;
+
+struct VertexConstructor {
+    color: epaint::Color32,
+}
+impl FillVertexConstructor<epaint::Vertex> for VertexConstructor {
+    fn new_vertex(&mut self, vertex: tessellation::FillVertex) -> epaint::Vertex {
+        let pos = egui::pos2(vertex.position().x, vertex.position().y);
+        epaint::Vertex { pos, uv: epaint::WHITE_UV, color: self.color }
+    }
+}
+
+impl StrokeVertexConstructor<epaint::Vertex> for VertexConstructor {
+    fn new_vertex(&mut self, vertex: tessellation::StrokeVertex) -> epaint::Vertex {
+        let pos = egui::pos2(vertex.position().x, vertex.position().y);
+        epaint::Vertex { pos, uv: epaint::WHITE_UV, color: self.color }
+    }
+}
+
+enum RenderOp {
+    Delete,
+    Paint(egui::Shape),
+    Transform(Transform),
+}
+
+pub struct Renderer {
+    mesh_cache: HashMap<String, egui::Shape>,
+    pub painter: Option<egui::Painter>,
+    dark_mode: bool,
+}
+
+impl Renderer {
+    pub fn new(elements_count: usize) -> Self {
+        Self { mesh_cache: HashMap::with_capacity(elements_count), painter: None, dark_mode: false }
+    }
+
+    pub fn render_svg(&mut self, ui: &mut egui::Ui, buffer: &mut Buffer, painter: egui::Painter) {
+        let mut elements = buffer.elements.clone();
+
+        self.painter = Some(painter.clone());
+        let dark_mode_changed = ui.visuals().dark_mode != self.dark_mode;
+        self.dark_mode = ui.visuals().dark_mode;
+
+        // todo: should avoid runing this on every frame, because the images are allocated once
```

Comment:
I assume this is okay to do later since the PR's ready for review

Reply:
yeah, not a heavy perf hit. I'm just scanning through the images and allocating textures to newly added ones. most of the frames, the images will have already been allocated and no duplicate texture will be reallocated. so this wastes an O(images) runtime.